numba/experimental/function_type.py

```diff
@@ -249,17 +285,48 @@ def lower_cast_function_type_to_function_type(
 def lower_cast_dispatcher_to_function_type(context, builder, fromty, toty, val):
     toty = toty.get_precise()
 
-    pyapi = context.get_python_api(builder)
+    sig = toty.signature
+    dispatcher = fromty.dispatcher
+    llvoidptr = context.get_value_type(types.voidptr)
     sfunc = cgutils.create_struct_proxy(toty)(context, builder)
+    # Always store the python function
+    sfunc.py_addr = builder.ptrtoint(val, llvoidptr)
+
+    # Attempt to compile the Dispatcher to the expected function type
+    try:
+        cres = dispatcher.get_compile_result(sig)
+    except errors.NumbaError:
+        cres = None
+
+    # If compilation is successful, we can by-pass using GIL to get the cfunc
+    if cres is not None:
+        # Declare cfunc in the current module
+        wrapper_name = cres.fndesc.llvm_cfunc_wrapper_name
+        llfnptr = context.get_value_type(toty.ftype)
+        llfnty = llfnptr.pointee
+        fn = cgutils.get_or_insert_function(
+            builder.module, llfnty, wrapper_name,
+        )
+        addr = builder.bitcast(fn, llvoidptr)
+        # Store the cfunc
+        sfunc.c_addr = addr
+        # Store the jit func
+        fn = context.declare_function(builder.module, cres.fndesc)
+        sfunc.jit_addr = builder.bitcast(fn, llvoidptr)
+        # Link-in the dispatcher library
+        context.active_code_library.add_linking_library(cres.library)
+
+    else:
+        # Use lower_get_wrapper_address() to get the cfunc
+        lower_get_wrapper_address
+        pyapi = context.get_python_api(builder)
+
+        gil_state = pyapi.gil_ensure()
+        addr = lower_get_wrapper_address(
+            context, builder, val, toty.signature,
+            failure_mode='return_exc')
+        sfunc.c_addr = pyapi.long_as_voidptr(addr)
```

Comment:
Should the `sfunc.jit_addr` explicitly be set to `NULL` on this path?

Reply:
It is zero-filled by default due to `create_struct_proxy()` which did it at: https://github.com/numba/numba/blob/ed19445d09694bdb82ee00e112c027a1685ee546/numba/core/cgutils.py#L130

Reply:
ah yes, thanks, that makes sense.